### PR TITLE
Correction of CollectionView to work properly with selected item colo…

### LINKED ISSUE #7790: [Bug] ItemsViewCell SelectedBackground shows Gray selected background even when content visual state has been overriden
### Description
**On iOS**
When using a CollectionView, overriding the selected visual states of the DataTemplate content the Gray SelectBackgroundView is shown and cannot be hidden.   A long press will show the Gray selectedbackground color, then the overridden one which looks odd.

Additionally, if the content doesn't fill the whole cell(such as using rounded corners) the gray background selection shows through on the unused space which looks bad(see below)

**On Android**
The same thing happens but it uses the system selection colour if the overriden background colour is Transparent.

                <DataTemplate>
                    <Grid Margin="10" BackgroundColor="Transparent" HeightRequest="200">
                        <VisualStateManager.VisualStateGroups>
                           <VisualStateGroupList>
                                <VisualStateGroup x:Name="CommonStates">
                                    <VisualState x:Name="Normal" />
                                    <VisualState x:Name="Selected">
                                        <VisualState.Setters>
                                            <Setter Property="BackgroundColor" Value="Transparent" />
                                        </VisualState.Setters>
                                    </VisualState>
                                </VisualStateGroup>
                            </VisualStateGroupList>
                        </VisualStateManager.VisualStateGroups>

The justification is that i want to handle the selection change within the visible content and there is no way to stop the full cell background colour(Gray on iOS and System colour on android) being shown and too leave selection to the templated content.

This is the code on iOS that is causing the problem

https://github.com/xamarin/Xamarin.Forms/blob/37c22eb1cb64afc5d62b9386a50bb6090446a65a/Xamarin.Forms.Platform.iOS/CollectionView/ItemsViewCell.cs#L16-L21

### Steps to Reproduce

1. Create a new XF project
2. Add a CollectionView, set to use an ItemSource and a DataTemplate and SelectionMode set to single
3. Add a Grid(with the visual state as above) as the root item object, then add a Frame as below in the ItemTemplate
  - Margin = "10"
  - BackgroundColor = "Red"
  - CornerRadius = "20"

### Expected Behavior
When the item is selected the unused background will not change colour
### Actual Behavior
The unused background shows either Gray(iOS) and system color on Android

You could provide a property of CollectionView to set the selected item colour, or have some other way of disabling it.  I could disable selection and add a tap gesture to the sub-views but that would be more work and would lose some of the benefits of CollectionView.

### Basic Information

- Version with issue:  4.2 
- Last known good version: n/a
- IDE: VS4M 2019 8.3
- Platform Target Frameworks: 
  - iOS:  13.0
  - Android:   
- Android Support Library Version: 
- Nuget Packages:
- Affected Devices:

### Screenshots

Screenshot 2019-10-03 at 09 42 47

### Reproduction Link

[CollectionViewTest.zip](https://github.com/xamarin/Xamarin.Forms/files/3685369/CollectionViewTest.zip)

### LINKED ISSUE #9590: [Bug] Default selection colors for platforms interfere with selected styles
### Description

CollectionView uses the default platform "selected" colors for selected items if nothing is specified for the "Selected" VisualState. 

Unfortunately, if the styling you want for your "Selected" state is transparent or offset, the platform's selection color shows through. 

The platform selection colors need to be removed/cleared if Selected state is specified.

### LINKED ISSUE #11421: [Bug] [UWP, iOS] Can't set transparent color on selected template item in CollectionView
### Description
Can't set transparent color on selected template item in CollectionView on iOS. Other colors work. As i found the issue with "selected" state in UWP is being resolved so maybe this is only an iOS issue
### Steps to Reproduce

1. Try the code from official sample docs: [*click*](https://docs.microsoft.com/ru-ru/xamarin/xamarin-forms/user-interface/collectionview/selection#change-selected-item-color) 
2. Set background color to Transparent 
3. See that the color is not being changed to transparent on iOS (uwp).

### Expected Behavior
Color is being set to Transparent on all platforms

### Actual Behavior
Color is not being set to Transparent on iOs (uwp)

### Workaround
Set background color as property of binded observablecollection item

### PR DESCRIPTION
**Pull request by ronymesquita**
### Description of Change ###

The key idea is to resolve the problem of CollectionView with selected item color on Android and iOS.

### Issues Resolved ### 

- fixes #7790
- fixes #9590
- fixes #11421

### API Changes ###
 
 None

### Platforms Affected ### 

- iOS
- Android

### Behavioral/Visual Changes ###

The platform will check if exists a `Visual State` with `Selected` and not use the default color. In this manner, the background color of the `Visual State` will not be overridden by Android. On iOS, the background color is cleaned when the `Selected` state exists. 
The default behavior still works.

#### After (Android)
android

#### After (iOS)
ios

### Testing Procedure ###
The test was manual.

### PR Checklist ###

- [ ] Targets the correct branch
- [ ] Tests are passing (or failures are unrelated)

### CODE REVIEW

**Comment by dnfadmin** [comment]
[CLA assistant check](https://cla.dotnetfoundation.org/xamarin/Xamarin.Forms?pullRequest=11604) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/xamarin/Xamarin.Forms?pullRequest=11604) before we can accept your contribution.<br/><br/>:x: ronymesquita [sign now](https://cla.dotnetfoundation.org/xamarin/Xamarin.Forms?pullRequest=11604)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/xamarin/Xamarin.Forms?pullRequest=11604) it.</sub>

**Comment by ronymesquita** [comment]
The code no check to see if that are more than one `Visual State` with `Selected` state on the page.

**Comment by ronymesquita** [comment]
The [SingleBoundSelection.xaml](https://github.com/xamarin/Xamarin.Forms/pull/11604/files#diff-92bb0a853f14e2626dd85a736dca9a06) file has been changed for testing purposes only.

**Comment by EmilAlipiev** [comment]
we need this asap. workaround with visualstates doesnt work in every case.

**Comment by ronymesquita** [comment]
Hello.

This pull request will be merged?

**Comment by developer9969** [comment]
Hi there is this going to be merged into 5?

**Comment by hartez** [comment]
This is a good start, but it will only work if the visual state for Selected is set in resources; it doesn't quite work if the state is set directly in the data template. Working on an update to correct that.

**Comment by hartez** [comment]
There, should work now even if the visual states are set in the template (see the changes to SingleBoundSelection.xaml to see what I mean).

**Review by hartez — APPROVED** [review]

**Comment by hartez** [comment]
Starting up the automated UI test run now, once those pass and we can get a second review we'll get this merged.

**Comment by ronymesquita** [comment]
Great!
:thumbsup:

**Review by rmarinho — CHANGES_REQUESTED** [review]
I think this needs rebase on top of 5.0. Also needs the correct formatting it's using spaces.

**Comment by developer9969** [comment]
@ronymesquita  @hartez  thank you for this fix , it will make a big difference to our apps as we use collection view all over the place and the selection item color was a big issue for our client

**Comment by hartez** [comment]
> I think this needs rebase on top of 5.0. Also needs the correct formatting it's using spaces.

Why do you think it needs a rebase? No conflicts.

**Comment by hartez** [comment]
We ran into some merging issues with this PR, so we had to move it to a new one: https://github.com/xamarin/Xamarin.Forms/pull/12651.

Same commits, just a new branch.

**Pull request closed without merging**
